File: MODULE 4/DAY 1/ProductSpringdemo/src/main/java/org/northernarc/productspringdemo/service/ProductService.java
package org.northernarc.productspringdemo.service;

import org.northernarc.productspringdemo.DTO.ProductRequestDTO;
import org.northernarc.productspringdemo.DTO.ProductResponseDTO;
import org.northernarc.productspringdemo.DTO.ProductUpdateDTO;
import org.springframework.security.access.prepost.PreAuthorize;

import java.util.List;

public interface ProductService {

    @PreAuthorize("hasRole('ADMIN')")
    ProductResponseDTO saveProduct(ProductRequestDTO productDTO);

    @PreAuthorize("hasAnyRole('USER','ADMIN')")
    List<ProductResponseDTO> getAllProducts();

    @PreAuthorize("hasAnyRole('USER','ADMIN')")
    ProductResponseDTO getProductById(Integer id);

    @PreAuthorize("hasRole('ADMIN')")
    ProductResponseDTO updateProduct(Integer id,
                                     ProductUpdateDTO productDTO);

    @PreAuthorize("hasRole('ADMIN')")
    void deleteProduct(Integer id);

    @PreAuthorize("hasAnyRole('USER','ADMIN')")
    List<ProductResponseDTO> getProductsByCategory(String category);

    @PreAuthorize("hasAnyRole('USER','ADMIN')")
    List<ProductResponseDTO> getProductsByBrand(String brand);

    @PreAuthorize("hasAnyRole('USER','ADMIN')")
    List<ProductResponseDTO> getProductsByName(String name);
}

File: MODULE 5/AngularPractice/jwtDemo/src/app/services/product-service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { inject } from '@angular/core';
import { ProductRequestDto, ProductUpdateDto, ProductResponseDto } from '../dto/ProductDTO';
import { Observable } from 'rxjs/internal/Observable';
@Injectable({ providedIn: 'root' })
export class ProductService {

  private readonly http = inject(HttpClient);
  private readonly baseUrl = '/api/ecom/product';

  getAllProducts(): Observable<ProductResponseDto[]> {
    return this.http.get<ProductResponseDto[]>(this.baseUrl);
  }

  getProductById(id: number): Observable<ProductResponseDto> {
    return this.http.get<ProductResponseDto>(`${this.baseUrl}/${id}`);
  }

  createProduct(dto: ProductRequestDto): Observable<ProductResponseDto> {
    return this.http.post<ProductResponseDto>(this.baseUrl, dto);
  }

  updateProduct(id: number, dto: ProductUpdateDto): Observable<ProductResponseDto> {
    return this.http.put<ProductResponseDto>(`${this.baseUrl}/${id}`, dto);
  }

  deleteProduct(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  getProductsByCategory(category: string): Observable<ProductResponseDto[]> {
    return this.http.get<ProductResponseDto[]>(`${this.baseUrl}/category/${encodeURIComponent(category)}`);
  }

  getProductsByBrand(brand: string): Observable<ProductResponseDto[]> {
    return this.http.get<ProductResponseDto[]>(`${this.baseUrl}/brand/${encodeURIComponent(brand)}`);
  }

  getProductsByName(name: string): Observable<ProductResponseDto[]> {
    return this.http.get<ProductResponseDto[]>(`${this.baseUrl}/name/${encodeURIComponent(name)}`);
  }
  
}

